scripts/sql/update.sql

```diff
@@ -0,0 +1,164 @@
+    ret text;
+  begin
+    execute s into ret;
+    return ret;
+  end;
+$$;
+
+
+create or replace function strip_comments(s text) 
+returns text
+language plpgsql
+as $$
+  declare
+    ret text;
+  begin
+    select regexp_replace( s, '^--[ \t-]+.*', '', 'gn') into ret;
+    return ret;
+  end;
+$$;
+
+
+create or replace function split_stmts(s text) 
+returns setof text
+language plpgsql
+as $$
+  declare
+    ret text;
+  begin
+    return query (select * from unnest(regexp_split_to_array(s, ';')));
+  end;
+$$;
+
+
+create or replace function file_read(filename text)
+returns text
+language plpgsql volatile
+as $$
+  declare
+    ret     text;
+    _oid    integer;
+  begin
+    select lo_import(filename) into _oid;
+    select encode(lo_get(_oid), 'escape') into ret;
+    perform lo_unlink(_oid);
+    return ret;
+  end
+$$;
+
+
+create or replace function runner(_path text, _basename text)
+returns void
+language plpgsql volatile
+as $$
+  declare _filename   text;
+  declare _exists     boolean;
+  declare _source_text text;
+  declare _source_md5 text;
+  declare _event_id   bigint;
+  declare _sql        text;
+  declare _trimmed_sql text;
+  declare _result     text;
+  declare _start      timestamptz;
+
+begin
+  -- have we previously successfully run the script?
+  select exists (
+    select 1
+      from event
+      where event.filename = _basename
+      -- and event.md5 = _md5
+    ) into _exists;
+
+  if _exists = false then
+    -- run it
+    raise notice 'running file: %', _basename;
+
+    select _path || '/' || _basename into _filename;
+    select file_read(_filename) into _source_text;
+    select md5(_source_text) into _source_md5;
+
+    insert
+      into event(filename, source, md5)
+      values (_basename, _source_text, _source_md5) 
+      returning id
+      into _event_id;
+
+    for _sql in select split_stmts(strip_comments(_source_text))
+    loop
+
+      select regexp_replace( _sql, '^\s*(.*)\s*$', '\1', 'g') into _trimmed_sql;
+      if _trimmed_sql <> '' then
+        raise notice 'running sql: %', _trimmed_sql;
+
+        select now() into _start;
+        select exec_returning_string(_sql) into _result;
+
+        insert into action(event_id, sql, result, start, finish)
+        values (_event_id, _trimmed_sql, _result, _start, now());
+
+      end if;
+    end loop;
+
+  else
+    -- report only
+    raise notice 'skipping: %', _filename;
+  end if;
+end 
+$$;
+
+
+begin;
+
+set search_path=aatams,public;
+
+create table if not exists event
+(
+  id        bigserial primary key,
+  filename  text not null,
+  source    text not null,
+  md5       text not null
+);
+
+create table if not exists action
+(
+  id        bigserial primary key,
+  event_id  bigint not null references event(id),
+  sql       text not null,
+  result    text not null,
+  start     timestamptz not null,
+  finish    timestamptz not null
+);
+
+
+select runner(:path, 'xavier-changes-2016_April/aatamsDB_queries_#123.sql');
```

Comment:
The functions are useful outside of the aatams context and so I don't think the updates belong there.  Although we could just wait until we need to use them elsewhere and do it then.   Running the create/replace functions every time you want to run an update bothers me - its unnecessary.  Its only required because you don't want to add it to chef/keep track of whether you've added the functions before.